### ISSUE TITLE
Allow creation of personal access tokens

### ISSUE THREAD

**Issue by syropian** [spotify/web-api#534]
Many OAuth implementations let a user create "personal access tokens", that allow them to access _their own_ data without going through an entire authentication flow. This would simplify many common things people want like fetching their currently playing track, controlling playback etc. Ideally they would be long-lived to avoid futzing with refresh tokens as well. Any plans to add this? 🙏

**Comment by hughrawlinson** [comment]
Thanks for the feature request @syropian - we'll consider it, but don't rely on it becoming available in the short term.

**Comment by sahandnayebaziz** [comment]
@hughrawlinson this would be a great addition. I got here following the trail of sudden 401's coming back from `https://api.spotify.com/v1/search`. I am using the API simply to find the ID of a track by name to give users links to listen to that song when they want to. Similar thing I am doing with Apple Music, where the search endpoint is open and doesn't require a token.

Hopefully you can make this easier to address use cases that don't have a real "user context"

**Comment by hughrawlinson** [comment]
> Hopefully you can make this easier to address use cases that don't have a real "user context"

@sahandnayebaziz this is already relatively straightforward using the [Client Credentials](https://developer.spotify.com/web-api/authorization-guide/#client-credentials-flow) flow. It only requires one HTTP request to get a Client Credentials access token, which expires once an hour. It does require your client secret though, so make sure to implement the flow server side so you don't leak the client secret!

I'll keep this open, as there could be _some_ use cases that may necessitate long-lived client credentials tokens - but the general advice at the moment is to implement client credentials (while keeping your client secret safe!) and request a new access token each time the previous one expires.

**Comment by sahandnayebaziz** [comment]
Ah! Client credentials flow. (Little bit of a confusing name, no? If it's intended for user server-side?)

I missed that. My mistake. Thank you @hughrawlinson :)

**Comment by hughrawlinson** [comment]
@sahandnayebaziz the client in this case refers to your Spotify application as created in the "My Apps" section of developer.spotify.com. The name comes from the [oauth 2.0 specification](https://tools.ietf.org/html/rfc6749#section-1.1).

**Comment by maccettura** [comment]
@hughrawlinson Maybe I am missing something, but would the "Client Credentials" flow allow the application to manage Player data?  My understanding was that it could not be used for "personal" data.  I basically need an easy way for one machine to manage the playing/pausing/queueing of tracks of a spotify player.

**Comment by NeoLegends** [comment]
@maccettura No, the client credentials flow can only be used for non-user data. If you want to manage user-data you'll have to go through the OAuth Client Credentials flow.

**Comment by NeoLegends** [comment]
> Many OAuth implementations let a user create "personal access tokens", that allow them to access their own data without going through an entire authentication flow. This would simplify many common things people want like fetching their currently playing track, controlling playback etc. Ideally they would be long-lived to avoid futzing with refresh tokens as well. Any plans to add this? 🙏

This would also simplify development and testing, since your half-finished tool (or your automated tests) wouldn't have to go through the full OAuth procedure.

**Comment by vaibhavhrt** [comment]
So to manage even my own playlists I have to go through this whole process.

**Comment by jscholes** [comment]
Yes, because otherwise the authentication process is undermined.  If you can take a shortcut into your account for API access purposes, so can an attacker.

It really isn't that arduous.  Just set up a one-time way to get access and refresh tokens.  You don't even need your software, library or whatever to be ready yet - just set the callback URI to something like httpbin.  When your browser redirects you back there, httpbin will show the auth code, request the tokens with CURL and you're done.

If Spotify were to implement this, they'd also have to think about what happens when you leave your personal token lying around and somebody gains access to it.  Given that for a long time they didn't even allow people to de-authorize apps from their account, it seems unlikely they'd bother.

**Comment by syropian** [comment]
Generally users can revoke or rotate any access tokens they have. Spotify data also isn’t particularly sensitive compared to something like GitHub who _does_ offer personal access tokens.

**Comment by vaibhavhrt** [comment]
Well I created a simple flask app for testing the app, isn't too much trouble. But a personal access token would be nice.
